### PR TITLE
Add a bookinfo-reviews-v3 example for multicluster

### PR DESCRIPTION
**Pull request by sdake**
This is intended to be used with route-rule-reviews-v3.yaml, coupled
with documentation to be submitted via PR to the github.istio.io repository
to demonstrate how to deploy bookinfo in a multicluster scenario.  This
file must be sidecar injected and run on the "remote" cluster independently
after bookinfo-reviews-v3 deployment is removed from the Istio control plane.

Documentation should hit the PR queue by EOD Friday.

While the multicluster solution now is imperfect, the general idea is to get
the basics documented and exampled so others in the community can eval it,
provide feedback, and hopefully improve the implementation incrementally.

This may or may not be a long lived file in the repository.  I am hopeful it
is not long-lived, but will be necessary to achieve the incremental development
approach described above.

### CODE REVIEW

**Comment by john-a-joyce** [comment]
Other than adding a comment this look sgood to me.

**Comment by sdake** [comment]
@frankbu this is similar in nature to the previous PR related to multicloud.

**Comment by frankbu** [comment]
Since this is probably temporary, maybe it would be better to just inline the yaml in the command that you provide in the doc, e.g.:
```
cat <<EOF | kubectl create -f -
apiVersion: v1
kind: Service
metadata:
  name: reviews
  labels:
    app: reviews
spec:
  ports:
  - port: 9080
    name: http
  selector:
    app: reviews
---
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: reviews-v3
spec:
  replicas: 1
  template:
    metadata:
      labels:
        app: reviews
        version: v3
    spec:
      containers:
      - name: reviews
        image: istio/examples-bookinfo-reviews-v3:1.5.0
        imagePullPolicy: IfNotPresent
        ports:
        - containerPort: 9080
EOF
```

**Comment by sdake** [comment]
@frankbu I had thought of that, although I don't think the bookinfo examples will be temporary - although I am hopeful we can remove them prior to 1.0, I can't guarantee it.

Putting big yamls in documentation even with cut and paste makes it difficult for developers to update and iterate the implementation as the documentation is in a separate repository.

File: samples/bookinfo/kube/bookinfo-reviews-v3.yaml [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+  name: reviews
+  labels:
+    app: reviews
+spec:
+  ports:
+  - port: 9080
+    name: http
+  selector:
+    app: reviews
+---
```
Comment (rshriram) [inline samples/bookinfo/kube/bookinfo-reviews-v3.yaml]: how will this work, since we already declared a service for reviews?

**Comment by rshriram** [comment]
While I am totally for adding this to our repo, I think it should wait to do so, until we have some CI tests to evaluate it. Until then, I am with @frankbu in that we can easily inline this YAML in the docs. It makes it much more easy for us to iterate on it. If there are simple bugs or things to fix, we can fix the docs within the same release, rather than trying to get another release out to fix these yamls. 

Besides, this thing is redeclaring reviews service - which kind of creates confusion/deviates from bookinfo.yaml. So, I am not sure how this plays out.

**Comment by sdake** [comment]
There is a CI PR for Multicluster here in need of review: https://github.com/istio/istio/pull/4240

This PR in question is is mostly for the docs though - the first PR for the documentation is here: https://github.com/istio/istio.github.io/pull/1139

I am not sure the CI PR has this integrated.  The general idea is to get all of these things integrated, but to do that (since multiple people are doing the work) the patches need to hit the repo.

I don't plan to add more samples to the samples directory after this one but would find it very helpful to not have to inline in the docs.

Cheers
-steve

**Comment by frankbu** [comment]
The bottom line is that all these `bookinfo-XXX.yaml` files are a mess. This one doesn't even conform to the naming convention of other `bookinfo-XXX-vX.yaml` files in that the other ones include just the Deployment, while this one also includes the corresponding Service. I think we need to clean this up.

Maybe we should have two folders `deployments` and `services`. In the `services` folder we could put one yaml file for each of the 4 services (e.g., `bookinfo-reviews.yaml`, `bookinfo-ratings.yaml`, etc) and in the `deployments` folder we put one file for each Deployment (e.g., `bookinfo-reviews-v1.yaml`, `bookinfo-reviews-v2.yaml`, etc). That way we can pick and choose whatever single resource we want to apply. In the `kube` directory we only keep `bookinfo.yaml`, a convenient yaml that has all the resources needed for the basic tasks in the doc (could even be generated by the build?).

cc @vadimeisenbergibm: I think this is something along the lines of what you were suggesting.

**Comment by vadimeisenbergibm** [comment]
@frankbu let me reopen my PR about Bookinfo restructuring https://github.com/istio/istio/pull/4042. Now 0.7 is out, we can get back to it.

**Comment by sdake** [comment]
@frankbu I see, what is proposed is a meta-solution to my micro-problem :)  I do agree with you that the bookinfo samples are a bit of a mess.  Your proposal appears tactically sound.  If @vadimeisenbergibm wants to take on that meta work item, that would be great.  I feel a bit like I don't have enough expertise in bookinfo specifically to make sure I don't break things with a huge reorg of the bookinfo samples.

If I should just abandon this PR, please let me know so I can move on to other PRs.

Alternatively if there is some other corrective action I can take to simplify this PR without interfering with @vadimeisenbergibm 's rework, please let me know.  I'll make that corrective action and push additional commits for merge.

Cheers
-steve

**Comment by tiswanso** [comment]
Just to update here regarding the pilot multicluster e2e test & CI work being done... #4240 is focused on getting the existing istio-pilot-e2e tests running in a multicluster setup.   So, it's deploying several instances of the test app used for istio-pilot-e2e.

A follow-on PR (dependent on #4240) will add additional tests that prove out more of the communication paths applicable specifically to the multicluster environment.  

So, there is no direct overlap with that work and the bookinfo example @sdake is adding content for other than relying on the same multicluster feature.

**Comment by vadimeisenbergibm** [comment]
@sdake I will do the restructuring of the bookinfo sample. Your PR will not interfere with my work, I will restructure the files of your PR together with other bookinfo files.

**Comment by sdake** [comment]
@vadimeisenbergibm cool - I had a look at your PR about an hour ago prior to the env wg meeting.  Looks pretty good and should meet the needs of a bookinfo multicluster sample document nicely.

Cheers
-steve

**Comment by sdake** [comment]
Closing in favor of solution by @vadimeisenbergibm in PR: https://github.com/istio/istio/pull/4042

**Pull request closed without merging**